dll/win32/shell32/folders/CRecycleBin.cpp

```diff
@@ -982,6 +982,91 @@ TRASH_TrashFile(LPCWSTR wszPath)
     return DeleteFileToRecycleBin(wszPath);
 }
 
+static void TRASH_PlayEmptyRecycleBinSound()
+{
+    LPCWSTR lpRegSubKey = L"AppEvents\\Schemes\\Apps\\Explorer\\EmptyRecycleBin";
+    HKEY hRegKey, hRegSnd;
+    LPWSTR lpRegVal, lpSndPath;
+    DWORD dwRegValType, dwRegValSize, dwSndPathLen;
+    LONG lRegStatus;
+
+    /* Open registry key */
+    lRegStatus = RegOpenKeyExW(HKEY_CURRENT_USER, lpRegSubKey, 0, KEY_QUERY_VALUE, &hRegKey);
+
+    if (lRegStatus != ERROR_SUCCESS)
+    {
+        return;
+    }
+
+    /* Open .Current */
+    lRegStatus = RegOpenKeyExW(hRegKey, L".Current", 0, KEY_QUERY_VALUE, &hRegSnd);
+
+    if (lRegStatus != ERROR_SUCCESS)
+    {
+        /* If fail then open .Default */
+        lRegStatus = RegOpenKeyExW(hRegKey, L".Default", 0, KEY_QUERY_VALUE, &hRegSnd);
+
+        RegCloseKey(hRegKey);
+
+        if (lRegStatus != ERROR_SUCCESS)
+        {
+            return;
+        }
+    }
+
+    RegCloseKey(hRegKey);
+
+    /* Get registry value size */
+    lRegStatus = RegQueryValueExW(hRegSnd, NULL, NULL, &dwRegValType, NULL, &dwRegValSize);
+
+    /* Check whether the type is valid */
+    if (lRegStatus != ERROR_SUCCESS || (dwRegValType != REG_SZ && dwRegValType != REG_EXPAND_SZ))
+    {
+        RegCloseKey(hRegSnd);
+        return;
+    }
+
+    /* Allocate buffer for registry value */
+    lpRegVal = (LPWSTR)HeapAlloc(GetProcessHeap(), 0, (SIZE_T)dwRegValSize);
```

Comment:
Ditto, this is c++ why not use `CComHeapPtr<WCHAR>` instead of manually handling that?

Reply:
It's depends on ATL.